src/v/iceberg/transform_utils.cc

```diff
@@ -45,4 +63,145 @@ value apply_transform(const value& source_val, const transform& transform) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const string_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const binary_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const fixed_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const decimal_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const uuid_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const auto& f) {
+              return std::make_optional<partition_spec_field_error>(fmt::format(
+                "Can not apply truncate transform to the filed {}", f));
+          });
+    }
+    std::optional<partition_spec_field_error>
+    operator()(const truncate_transform&) {
+        return ss::visit(
+          field_type,
+          [this](const auto&) {
+              return std::make_optional<partition_spec_field_error>(fmt::format(
+                "Can not apply truncate transform to the filed {}",
+                field_type));
+          },
+          [](const int_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const long_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const string_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const binary_type&) {
+              return std::optional<partition_spec_field_error>{};
+          },
+          [](const decimal_type&) {
+              return std::optional<partition_spec_field_error>{};
+          });
+    }
+    std::optional<partition_spec_field_error>
+    operator()(const year_transform& tr) {
+        return validate_date_time_transform(tr);
+    }
+    std::optional<partition_spec_field_error>
+    operator()(const month_transform& tr) {
+        return validate_date_time_transform(tr);
+    }
+    std::optional<partition_spec_field_error>
+    operator()(const day_transform& tr) {
+        return validate_date_time_transform(tr);
+    }
+    std::optional<partition_spec_field_error>
+    operator()(const hour_transform&) {
+        if (
+          std::holds_alternative<timestamp_type>(field_type)
+          || std::holds_alternative<timestamptz_type>(field_type)) {
+            return std::nullopt;
+        }
+        return partition_spec_field_error(fmt::format(
+          "Can not apply hour transform to non timestamp field: {}",
+          field_type));
+    }
+    std::optional<partition_spec_field_error>
+    operator()(const void_transform&) {
+        return std::nullopt;
+    }
+
+private:
+    template<typename T>
+    std::optional<partition_spec_field_error>
+    validate_date_time_transform(const T& transform_type) {
```

Comment:
nit: maybe `validate_date_timestamp_transform` given it returns null for the time type?